[p2p] Add Peer struct for per-peer data in net processing

jnewbery opened the pull request:
We currently have two structures for per-peer data:

- `CNode` in net, which should just contain connection layer data (eg socket, send/recv buffers, etc), but currently also contains some application layer data (eg tx/block inventory).
- `CNodeState` in net processing, which contains p2p application layer data, but requires cs_main to be locked for access.

This PR adds a third struct `Peer`, which is for p2p application layer data, and doesn't require cs_main. Eventually all application layer data from `CNode` should be moved to `Peer`, and any data that doesn't strictly require cs_main should be moved from `CNodeState` to `Peer` (probably all of `CNodeState` eventually).

`Peer` objects are stored as shared pointers in a net processing global map `g_peer_map`, which is protected by `g_peer_mutex`. To use a `Peer` object, `g_peer_mutex` is locked, a copy of the shared pointer is taken, and the lock is released. Individual members of `Peer` are protected by different mutexes that guard related data. The lifetime of the `Peer` object is managed by the shared_ptr refcount.

This PR adds the `Peer` object and moves the misbehaving data from `CNodeState` to `Peer`. This allows us to immediately remove 15 `LOCK(cs_main)` instances.

For more motivation see #19398

jnewbery committed:
[net processing] Add Peer

Peer is a struct for holding per-peer data. This structure is not
protected by cs_main since it does not contain validation-critical data.

jnewbery committed:
[net processing] Move misbehavior tracking state to Peer

Misbehavior tracking state is now contained in Peer instead of
CNode. It is no longer guarded by cs_main, but instead by a
dedicated m_misbehavior_mutex lock.

This allows us to remove 14 cs_main locks from net_processing.

jnewbery committed:
scripted-diff: rename misbehavior members

-BEGIN VERIFY SCRIPT-
sed -i 's/nMisbehavior/m_misbehavior_score/g' src/net_processing.cpp src/net_processing.h src/rpc/net.cpp src/qt/rpcconsole.cpp
-END VERIFY SCRIPT-

jnewbery commented:
Thank you @MarcoFalke @promag @jonatack @dongcarl @ariard @theuni @fjahr @hebasto @troygiorshev for reviewing the prerequisite PRs! This is the first PR in the cs_main_split sequence where we start getting benefit by reducing cs_main scope in net processing. It also introduces the new structure and locking model, so it's the PR that requires the most conceptual/approach review in the sequence. If this gets merged, then it's a long procession of PRs to slowly move data into the new `Peer` object.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -468,6 +468,44 @@ static CNodeState *State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
+ */
+struct Peer {
+    /** Same id as the CNode object for this peer */
+    NodeId m_id;
```
```suggestion
    const NodeId m_id;
```

in commit d60a16bb3f58fcb9422c277256b3c3a31cab3b2f

Any reason to not make this const? Having this mutable would violate the assumption that the id is unique per peer.

jnewbery replied:
You're right. This should be const. Fixed

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -468,6 +468,44 @@ static CNodeState *State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
+ * cs_main since it does not contain validation-critical data.
+ *
+ * Memory is owned by shared pointers and this object is destructed when
+ * the refcount drops to zero (in FinalizeNode()).
```
nit in commit d60a16bb3f58fcb9422c277256b3c3a31cab3b2f
```suggestion
 * the refcount drops to zero.
```

FinalizeNode only decreases the refcount by one. It can still be larger than one (as long as someone is holding a PeerRef, e.g. `getpeerinfo` or similar)

jnewbery replied:
Fixed

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -468,6 +468,44 @@ static CNodeState *State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
+    PeerRef ret;
+    LOCK(g_peer_mutex);
+    auto it = g_peer_map.find(id);
+    if (it != g_peer_map.end()) ret = it->second;
+    return ret;
```
nit in same commit:

Can be written with one less line ;)

```suggestion
    if (it != g_peer_map.end()) return it->second;
    return {}; // ( or return nullptr; )
```

troygiorshev replied:
This may be a rare case where a ternary makes things clearer.
```suggestion
    return (it != g_peer_map.end()) ? it->second : nullptr;
```

jnewbery replied:
Done with Troy's ternary. Thanks!

maflcko reviewed (commented):
Approach ACK 3fa130da3ad7960f9ece4d34ff6e108b64b6f957, only some style-nits

troygiorshev reviewed (commented):
tACK 3fa130da3ad7960f9ece4d34ff6e108b64b6f957 with some questions and suggestions.

- Is the plan to lock every member of `Peer` behind its own mutex?  Why not give each `Peer` its own mutex instead?
- `Misbehaving`'s `cs_main` locks can be removed from` denialofservice_tests.cpp`: around [here](https://github.com/jnewbery/bitcoin/blob/3fa130da3ad7960f9ece4d34ff6e108b64b6f957/src/test/denialofservice_tests.cpp#L236).  I looked and couldn't find any further places.
- `clang-format` complains a bit on each commit.

MarcoFalke commented on src/net_processing.cpp:
```diff
@@ -468,6 +468,44 @@ static CNodeState *State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
+ * their own locks.
+ */
+Mutex g_peer_mutex;
+static std::map<NodeId, PeerRef> g_peer_map GUARDED_BY(g_peer_mutex);
```
in commit d60a16b

Is there any reason to make this a global? This map gets populated in `PeerLogicValidation::InitializeNode`, which is called by connman. `PeerLogicValidation` keeps a pointer to this connman, so it would seem natural to me that it also keeps track of the peer map, e.g. in a (private) member.

jnewbery replied:
This was mostly just following the convention of net_processing having its data stored in globals. I've had a look at what would be involved in moving `g_peer_map` and `GetPeer()` to be members of `PeerLogicValidation`. The main problem is that in future commits I want to be able to call `GetPeer()` from functions all across net_processing, and many of those aren't member functions of PLV. That means I'd need to either pass a reference to PLV down to those functions, or make them members of PLV.

Doing that (and moving more of the state currently stored as globals into PLV) seems like a reasonable thing to do, but feels a bit separate from the immediate goal of moving data into `Peer` so I'm inclined to leave it for later. What do you think?

jnewbery replied:
If I can get buy-in here from everyone else to move a bunch of the static functions from net_processing into PLV, then I'm very happy to move `g_peer_map` into PLV and move those functions as needed in future PRs.

theuni replied:
+1 on this. No preference whether it happens here or in a follow-up, but I agree PLV is a better home.

jnewbery replied:
I started trying to do this, but it's an enormous change. `Misbehaving()` fetches the `PeerRef`, and `Misbehaving()` is called all over net_processing. Let's defer this change to its own PR (probably a series of PRs since it touches almost all of the functions in net_processing.

jnewbery replied:
> I wonder if the ~CNetProcessing destructor should be cleaning up this global map, if this isn't moved into PLV? It looks like we don't clean up mapNodeState there, but that seems like an oversight -- unless there's some good reason I'm missing?

All nodes are cleaned up by `CConnman` during the Shutdown sequence (`Shutdown()` in init.cpp calls `CConnman.StopNodes()`, which calls `FinalizeNode()` for each node). When we come to destruct the `PeerLogicValidation` object both maps should already be empty.

I could add a `PeerLogicValidation` dtor which asserts these maps are empty, but that seems a bit orthogonal to this PR (if we introduce a bug that means the nodes are finalized during shutdown, then they'll exist in both the CNodeState map and the Peer map.

jnewbery replied:
@MarcoFalke @theuni @sdaftuar - I've made a branch that puts peer_map inside PeerLogicValidation here: https://github.com/jnewbery/bitcoin/tree/2020-08-peer-plv2. It builds on top of (merged) #19704 and (open) #19791. Let me know what you think. If that 19791 gets merged first, I'll switch this PR to use that branch. If not, I'm happy to do the follow-up work to move g_peer_map into PeerLogicValidation later.

jnewbery replied:
PR: #19910

jnewbery committed:
[net processing] Remove CNodeState.name

This has been unused since logging peer IPs was removed from
Misbehaving() in a8865f8b.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -270,12 +270,6 @@ struct CNodeState {
-    int nMisbehavior;
-    //! Whether this peer should be disconnected and marked as discouraged (unless it has the noban permission).
-    bool m_should_discourage;
-    //! String name of this peer (debugging/logging purposes).
-    const std::string name;
```
It absolutely should. When I wrote the original branch, I moved `name` to `Peer` as well, but since then, I removed all uses of `name` in #19472 and #19583 but failed to remove it as a member. I've now separated its removal into a separate commit.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -468,6 +459,51 @@ static CNodeState *State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
+
+/** Get a shared pointer to the Peer object.
+ *  May return nullptr if the Peer object can't be found. */
+static PeerRef GetPeer(NodeId id) {
```
ok, changed to `GetPeerRef()`

jnewbery commented on src/net_processing.cpp:
```diff
@@ -468,6 +459,51 @@ static CNodeState *State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
 }
 
+/**
+ * Data stucture for an individual peer. This struct is not protected by
```
Fixed

jnewbery commented:
> Is the plan to lock every member of Peer behind its own mutex? Why not give each Peer its own mutex instead?

Currently the plan is to have related members behind a single mutex, eg all tx relay fields behind the same mutex. The main reason is that that model maps most closely to what exists already (look at the various locks inside CNode). Once everything has moved to `Peer` then it should be much easier to change the locking model, but I didn't want to add a debate about what the ideal locking model is to this project.
 
> Misbehaving's cs_main locks can be removed from denialofservice_tests.cpp: around here. I looked and couldn't find any further places.

Fixed. Thanks!

> clang-format complains a bit on each commit.

Can you expand a bit. What am I supposed to be looking for here?

jnewbery commented:
> clang-format complains a bit on each commit.

ok, I've taken clang-format's suggestions, except one that joins an initializer list onto the same line as the function and parameter list (which makes a really long line), and one which would break the scripted-diff linter.

jonatack commented:
@jnewbery it looks like there is an incomplete rebase-in-progress in the last commit 123bba3b03a6b0f9d791758cb5dfc2e5468bf193; will review on next push.

jnewbery commented:
@jonatack oops. Fixed. Thanks!

jonatack commented on src/net_processing.cpp:
```diff
@@ -468,6 +459,50 @@ static CNodeState *State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
+ */
+struct Peer {
+    /** Same id as the CNode object for this peer */
+    const NodeId m_id;
```
337528b ISTM this type is an `int64_t` defined in net.h; should we initialize a default here for safety, like for the other members `m_misbehavior_score{0}` and `m_should_discourage{false}` added in 36ac41a

jnewbery replied:
Sure. This gets set in the initializer list of the only ctor, but there's no harm in adding this.

MarcoFalke replied:
It is impossible to not initialize a const member, so I'd prefer to keep it as is, but not strong opinion

jonatack commented on src/net_processing.cpp:
```diff
@@ -867,6 +907,13 @@ void PeerLogicValidation::ReattemptInitialBroadcast(CScheduler& scheduler) const
 }
 
 void PeerLogicValidation::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTime) {
+    int misbehavior;
```
36ac41a seems safer to initialize `misbehavior` with a value, even if set ATM a few lines further down

jnewbery replied:
ok. Done.

jonatack commented on src/net_processing.cpp:
```diff
@@ -867,6 +907,13 @@ void PeerLogicValidation::ReattemptInitialBroadcast(CScheduler& scheduler) const
 void PeerLogicValidation::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTime) {
+    int misbehavior;
+    {
+        PeerRef peer = GetPeerRef(nodeid);
```
36ac41a should we add an `if (peer == nullptr) return false;` check here as well

jnewbery replied:
I've added an assert, like for CNodeState. The Peer object should always exist when we try to remove the peer.

jonatack commented on src/net_processing.cpp:
```diff
@@ -3720,15 +3754,15 @@ void ProcessMessage(
 bool PeerLogicValidation::MaybeDiscourageAndDisconnect(CNode& pnode)
 {
     const NodeId peer_id{pnode.GetId()};
+    PeerRef peer = GetPeerRef(peer_id);
```
36ac41a

- perhaps this line inside the block like the three cases above

- should we add an `if (peer == nullptr) return false;` check here as well?

MarcoFalke replied:
A reference to the node exists (`CNode& pnode`) so it should be impossible for the peer ref to be deleted already

jnewbery replied:
> perhaps this line inside the block like the three cases above

I don't think this is required. The block is to limit the scope of the lock. Putting the `peer` variable in there is fine, but I don't think it's consistent with project style.

The block in `GetNodeStateStats()` was left over from a previous iteration of this branch. I've removed it. The block in `FinalizeNode()` is to force the `PeerRef` object to go out of scope, which will generally cause the underlying `Peer` object to be destructed, since that's usually the last reference.

> should we add an if (peer == nullptr) return false; check here as well?

Yes we should. Thank you!

jonatack commented on src/net_processing.cpp:
```diff
@@ -840,7 +875,12 @@ void PeerLogicValidation::InitializeNode(CNode *pnode) {
+    {
+        PeerRef peer = std::make_shared<Peer>(nodeid);
+        LOCK(g_peer_mutex);
+        g_peer_map.insert(std::make_pair(nodeid, std::move(peer)));
```
337528b not sure here, `s/insert/emplace/` to construct rather than copy?

jnewbery replied:
ok, I've changed this to `emplace_hint` since the new peers are always added to the end of the map (same as for CNodeStates in mapNodeState)

jonatack replied:
Interesting, TIL thanks, and there were only two `emplace_hint` statements in the codebase, one of which is in the same function. From https://en.cppreference.com/w/cpp/container/map/emplace_hint it looks like a good choice for lower complexity and better perf in this case.

jonatack reviewed (commented):
ACK afe3604f5e1a7eae5c38543147f55900569c9908

Some questions below, feel free to ignore.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -424,9 +422,8 @@ struct CNodeState {
-        address(addrIn), name(std::move(addrNameIn)), m_is_inbound(is_inbound),
-        m_is_manual_connection (is_manual)
+    CNodeState(CAddress addrIn, bool is_inbound, bool is_manual) :
+        address(addrIn), m_is_inbound(is_inbound), m_is_manual_connection (is_manual)
```
ah! Thank you!

jnewbery commented:
Thanks for the review @jonatack and @MarcoFalke . All your latest review comments are addressed.

troygiorshev commented:
reACK 951093884be4b6e48d2850dfc667c298f85afe9c

laanwj commented:
Code review ACK 951093884be4b6e48d2850dfc667c298f85afe9c

@theuni As you made the `CNode` and `CNodeState` separation originally can you take a look here please.

jonatack commented:
re-ACK 951093884be4b6e48d2850dfc667c298f85afe9c per `git diff afe3604 9510938`

fjahr commented:
Code review ACK 951093884be4b6e48d2850dfc667c298f85afe9c

Still testing a bit but everything looks good!

jnewbery commented:
Note for maintainers: please don't merge this yet. It has 4 ACKs, but going down this path is quite an important design decision. I'd like @theuni, @sipa or @sdaftuar to weight in before we move forward.

ariard reviewed (commented):
Approach ACK 9510938

I really like the approach to gather members by logical usage and progressively swap their cs_main guard by a new per-structure one.  I thought at first, it would add new lock dependencies if we moved everything from `CNodeState` directly into `Peer` as a quick read of related issue description made it thought but in fact no. 

Perhaps #19398 should laid out more the lock refactoring strategy and compare with any concurrent one (wasn't this discussed on IRC a while ago ?). It may help to take a decision here.

I glanced back on #18963, it adds also a `CPeerState`, aiming also to dry-up `CNodeState` but as branch tip was only used to signal end of _validation_ processing back to source peer and resume its _message_ processing. I think this current approach is somehow concurrent to #18963 but not incompatible with its end-goal of introducing asynchronous headers/block processing. Once this refactoring is done, it would be easier to reason on adding a new async "validation" thread, still challenging but easier.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -479,6 +473,13 @@ struct Peer {
     const NodeId m_id{0};
 
+    /** Protects misbehavior data members */
+    Mutex m_misbehavior_mutex;
```
I think it's easier to reason about having both misbehavior fields (`m_misbehavior_score` and `m_should_discourage`) under the same mutex since they're so closely related.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -479,6 +473,13 @@ struct Peer {
+    Mutex m_misbehavior_mutex;
+    /** Accumulated misbehavior score for this peer */
+    int nMisbehavior GUARDED_BY(m_misbehavior_mutex){0};
+    /** Whether this peer should be disconnected and banned (unless it has the noban permission) */
```
Thanks. This was a bad rebase. Now fixed.

sdaftuar reviewed (approved):
Seems fine.  Had a few questions but nothing that should be a blocker if you ignore. I'm not strongly wedded to the design decisions here though, it's not obvious to me that going down this road of having fine-grain, per-peer locks really adds anything, compared to (say) having just one lock for a whole Peer object (which maybe is even held the whole time you're using it, for simplicity).  But nothing here seems broken, and I don't see deadlock risks or anything for how this is being used now, so this seems okay to me.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -465,6 +465,43 @@ static CNodeState *State(NodeId pnode) EXCLUSIVE_LOCKS_REQUIRED(cs_main) {
+
+/** Get a shared pointer to the Peer object.
+ *  May return nullptr if the Peer object can't be found. */
+static PeerRef GetPeerRef(NodeId id)
```
Yes, that's possible. Do you think the comment on the map needs expanding? It currently says "Once a shared pointer reference is taken, the lock may be released. Individual fields are protected by their own locks."

jnewbery commented:
Thanks for the review @sdaftuar!

> I'm not strongly wedded to the design decisions here though, it's not obvious to me that going down this road of having fine-grain, per-peer locks really adds anything, compared to (say) having just one lock for a whole Peer object (which maybe is even held the whole time you're using it, for simplicity).

I'm not strongly wedded to it either. My reason for doing it this way is that it most closely reflects the locking design in CNode, so the future PRs that move the remaining data to Peer should be easier to review. Once they're all in Peer, moving from fine-grained mutexes to coarser-grained mutexes should also be an easy change if it's desired, but moving the other way requires a bit more thought. So in summary: this is easier and leaves more flexibility later.

jnewbery commented on src/net_processing.cpp:
```diff
@@ -479,6 +473,13 @@ struct Peer {
     /** Same id as the CNode object for this peer */
     const NodeId m_id{0};
 
+    /** Protects misbehavior data members */
```
🇺🇸

jnewbery commented on src/net_processing.cpp:
```diff
@@ -906,7 +907,11 @@ void PeerLogicValidation::ReattemptInitialBroadcast(CScheduler& scheduler) const
     {
+        PeerRef peer = GetPeerRef(nodeid);
+        assert(peer != nullptr);
+        misbehavior = WITH_LOCK(peer->m_misbehavior_mutex, return peer->nMisbehavior);
```
For functions that access multiple fields under the same lock, we'll need to hold the lock externally to the struct anyway (see `Misbehaving()` for example, which accesses both `m_misbehavior_score` and `m_should_discourage` while holding `m_misbehavior_score`). So I could add setters/getters which handle the locking, but the locks would still need to be public and it wouldn't help with encapsulating `Peer`.

I like that `Peer` is a struct with all public members and no internal logic, and I also think the `WITH_LOCK` macro is pretty nifty so don't mind this pattern. If others also have a strong distaste for the WITH_LOCK pattern, I can look at adding getter methods.

troygiorshev replied:
I'm fond of WITH_LOCK.  Especially if we're going the way of many individual locks, it's much quicker this way to see what's locking what.

This does give me flashbacks to Java though... `Button button = new Button("button");`

theuni replied:
@sdaftuar This was my high-level feedback as well. I strongly prefer encapsulation as it's really really hard to screw up encapsulated locking, but it's a burden when multiple operations could otherwise be handled with a single lock. But as this work/PR is intended to break up locking that has grown unwieldy, I can see why @jnewbery went for this approach.

Also, just in terms of evolution of thinking... I have always preferred encapsulated locks because of the security guarantees. But I'm less against exposed mutexes these days due to our annotations and sanitizers.

jnewbery replied:
It seems like there are two different points here:

- "This is a mouthful" (@sdaftuar) seems like a stylistic comment. I actually like the fact that the WITH_LOCK macro is explicit about the lock it's taking.
- "I strongly prefer encapsulation as it's really really hard to screw up encapsulated locking" (@theuni) is a comment about locking design. Cory, correct me if I'm wrong, but I think your distaste here is informed by how cs_main has been (mis)used in the past, where it's locked the mutex for huge scopes unnecessarily. With lock annotations and fine-grained mutexes, I don't see a problem with exposing those mutexes outside the class/struct.

theuni replied:
My primary concern is that 5 years from now someone changes some code that interacts with a Peer and forgets to grab a lock. With encapsulated locking, that _can't_ happen.

But I'm now satisfied that even without the encapsulation, we have safeguards to prevent that.

theuni commented on src/net_processing.cpp:
```diff
@@ -906,7 +907,11 @@ void PeerLogicValidation::ReattemptInitialBroadcast(CScheduler& scheduler) const
+    int misbehavior{0};
     {
+        PeerRef peer = GetPeerRef(nodeid);
+        assert(peer != nullptr);
```
This is mostly a question for future stuff that gets moved in but... What if the refcount is more than 1 here? Do we continue tearing it down?

As a contrived (bad) example: what if someone bumps the misbehavior score after it's been evaluated here?

jnewbery replied:
We'll remove the `Peer` shared ptr from `g_peer_map`, but we don't destruct the object until the refcount drops to 0. Your contrived example is fine. There would have to be another thread currently in `Misbehaving()` holding a shared ptr to the `Peer` object. We'd remove it from the map, the other thread would increment the misbehaving score, and then the `Peer` object would be destructed when that thread left the `Misbehaving()` function.

I've taken a look at the future commits in https://github.com/jnewbery/bitcoin/tree/2020-06-cs-main-split and I can't see any problems with this style of teardown.

theuni replied:
My point here was that calls to ```Misbehaving()``` might go unobserved, because the score may have already been evaluated before the other thread bumped it. This is not possible in with the current behavior because of cs_main: once we're in ```FinalizeNode```, we're guaranteed that no other thread is accessing it (unless they've cached the ```CNodeState*```, which would be a bug).

Obviously we don't have any threads doing that currently, so I'm wondering if we want to continue to enforce that invariant with a:
```c++
// This isn't a guarantee but it's close enough
assert(peer.use_count() <= 1);
```

I'm not too bothered if there's a rare misbehaving bump that gets missed. My concern is that Peer's destructor may eventually gain some functionality, and we'll want to know if we need to keep track of which thread deletes it.

jnewbery replied:
Thanks @theuni. This is a very good point. I'll add a comment about not relying on this being the final state of `Peer` in a future commit.

theuni commented on src/net_processing.cpp:
```diff
@@ -864,6 +906,10 @@ void PeerLogicValidation::ReattemptInitialBroadcast(CScheduler& scheduler) const
 void PeerLogicValidation::FinalizeNode(NodeId nodeid, bool& fUpdateConnectionTime) {
+    {
+        LOCK(g_peer_mutex);
+        g_peer_map.erase(nodeid);
```
Should this and ```mapNodeState.erase(nodeid)``` happen at the same time, under the same cs_main lock?

What are the consequences of them being out of sync? (Still reachable by ```State()``` but not by ```GetPeerRef()```)

jnewbery replied:
No bad consequences, but I've moved the Peer cleanup underneath the cs_main lock anyway to make this simpler to think about.

theuni reviewed (commented):
Left a few comments/questions. Conceptually I'm nervous about a shared_ptr's lifetime outliving its home in g_peer_map.

jnewbery commented:
Rebased (https://github.com/bitcoin/bitcoin/compare/8fe28ceb126dd3206f04f426aef8d067f538ec3f..3c56d2385e195d3de587df95f64ce97fc727b1a3)

jnewbery commented:
Thanks everyone for review. I've rebased and addressed all comments from @sdaftuar and @theuni . This is ready for re-review.

jnewbery commented:
@troygiorshev @laanwj @fjahr @jonatack this PR is now rebased and updated to address @sdaftuar and @theuni review comments. Do you mind re-ACKing?

troygiorshev commented:
reACK 8e35bf59062b3a823182588e0bf809b3367c2be0 via `git range-diff master 9510938 8e35bf5`

Only minor changes.

And, because I don't think I've seen it yet, ACK on the name of `g_peer_mutex`.  Usually our locks have a `cs` prefix.  I don't recall seeing a conflict like this before, our convention suggests this should be named something like `g_cs_peer_map` or `cs_peer_map`.  I like the current name better.  (In fact, in general I prefer the `mutex` suffix, leaving the prefix open to only `g` or `m`.)

promag commented:
Concept ACK.

theuni commented:
ACK 8e35bf59062b3a823182588e0bf809b3367c2be0.

I think my [comment](https://github.com/bitcoin/bitcoin/pull/19607#discussion_r470173964) here is important for reviewers to keep in mind here and for future work on ```Peer```, but I don't see any issues introduced as part of this PR.

jonatack reviewed (commented):
ACK 8e35bf59062b3a823182588e0bf809b3367c2be0 keeping in mind Cory's comment (https://github.com/bitcoin/bitcoin/pull/19607#discussion_r470173964) for the follow-up

sipa commented:
Concept ACK.
